### PR TITLE
Bug 1113094 — CLI Tools for OPS to get redis cluster status

### PR DESCRIPTION
**Pull request by Natim**
https://bugzilla.mozilla.org/show_bug.cgi?id=1113094

 - [x] Expiration: Gives back the number of keys that will expires with the size freed each day.
 - [x] Get number of users (detailed: active/inactive)
 - [x] purge data > 15 days (based on the idletime)
 - [x] Get the state of a hawk auth session (from the HawkId): anonymous, authenticated and oauth enabled;
 - [x] Get the average number of rooms per user;
 - [x] Get the average number of call-urls per user;
 - [x] Get the average number of calls per user;

### CODE REVIEW

**Comment by Natim** [comment]
Note that for this scripts we should avoir using the `TTL` command because it updates the `IDLETIME` that we need for some commands.

I am looking forward for https://github.com/antirez/redis/pull/2090

File: tools/active_unactive_users.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    options.port,
+    options.host,
+    options.options
+  );
+  if (options.db) client.select(options.db);
+
+  keysInformation(client, 'hawk.*', function(err, keysInfo) {
+    if (err) throw err;
+
+     console.log("Processing " + keysInfo.length + " keys");
```
Comment (leplatrem) [inline tools/active_unactive_users.js]: indent

File: tools/active_unactive_users.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    var monthly = 0;
+
+    keysInfo.forEach(function(key) {
+      var lruDate = new Date(Date.now() - key.lru_seconds_idle * 1000).getTime();
+      var now = Date.now();
+
+      var delta = (now - lruDate) / 1000;
+      var aDay = 3600 * 24;
+      var twoWeeks = 3600 * 24 * 7 * 2;
+      var aMonth = 3600 * 24 * 30;
```
Comment (leplatrem) [inline tools/active_unactive_users.js]: you can use uppercase constants like you did in remove_old_keys.js

**Comment by leplatrem** [comment]
Personnally, I would rename the command files so that they start with an action verb, what do you think ?

**Comment by Natim** [comment]
All with get_ ?

**Comment by leplatrem** [comment]
Maybe, yes. It sounds obvious, but when someones lists all available commands, there is no confusion that those are harmless (comparing to ``purge_``, ``remove_``, ``send_``...)

**Comment by Natim** [comment]
I have updated the PR and prefixed all commands that update the IDLETIME with an underscore.

File: tools/README.rst [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+Some of this tool are based on the IDLETIME value:
+
+ - ``get_active_unactive_users.js``
+ - ``get_expiration_estimate.js``
+ - ``remove_old_keys.js``
+
+The use of one of the first list's command will override the result of
+the second list's commands. To prevent errorneous usage, commands
+based on TTL are their name prefixed with an ``_``.
```
Comment (ametaireau) [inline tools/README.rst]: have their names

File: tools/README.rst [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+
+Active and Unactive users
```
Comment (ametaireau) [inline tools/README.rst]: Inactive

File: tools/README.rst [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+
+
+Keys expirations
+----------------
+
+- ``_get_expiration.js``
+- ``_graph_expiration.sh``
+- ``get_expiration_estimate.js``
+
+Each redis keys have got a Time-To-Live so we know when it would exipre.
```
Comment (ametaireau) [inline tools/README.rst]: "Each redis key has a"

File: tools/README.rst [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+
+Also because ``get_expiration_estimate`` is based on the IDLETIME, if you
+run it after ``_get_expiration.js`` all keys will have the same expiration
+date in the average time.
+
+
+Impacted users by the FxA bug
+-----------------------------
+
+- ``_get_fxa_impacted.js``
```
Comment (ametaireau) [inline tools/README.rst]: We probably should prefix this command with the bug number.

File: tools/README.rst [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+
+Providing an HawkId::
+
+    $ node get_hawk_user_info.js 88d5a28f545bb406ddc6c6a5276cbfe0aa10fdba425f4808e2d6c3acdbfdaeda
+    Trying with HawkIdHmac: de9cd5c5ded9e2df982723d96361f56c0d72c936dc177cbff1f147bac1445f63
+    { anonymous: false, userId: 'foobar@example.com' }
+
+Providing an HawkIdHmac::
+
+    $ nodeget_hawk_user_info.js de9cd5c5ded9e2df982723d96361f56c0d72c936dc177cbff1f147bac1445f63
```
Comment (ametaireau) [inline tools/README.rst]: there is a missing space here.

File: tools/README.rst [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+    userRooms.b8ae434636685b6d31c0b0efb96e649bd67c33c1c3fa9a23caaf3aaf804cfdd9
+    userRooms.494e14e5f507317b7392eafb3ca2a2372bd61a5735dbc06d9d70abe74b7d1d57
+
+    rooms.*: 	0
+
+
+Remove OLD keys
+---------------
+
+- ``remove_old_keys.js``
```
Comment (ametaireau) [inline tools/README.rst]: Can we specify a duration here? That could be useful to remove keys older than, for instance, the duration of a session.

File: tools/_get_expiration.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
+var conf = require('../loop/config').conf;
+var async = require('async');
+var redis = require("redis");
+
+var storage = conf.get("storage");
+
+var args = process.argv.slice(2);
+
+var verbose = args.indexOf('--verbose') !== -1;
```
Comment (ametaireau) [inline tools/_get_expiration.js]: Should we use something like https://www.npmjs.com/package/argparse to ease arguments for the CLI ?

File: tools/get_active_unactive_users.js [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  if (options.db) client.select(options.db);
+
+  keysInformation(client, 'hawk.*', function(err, keysInfo) {
+    if (err) throw err;
+
+    console.log("Processing " + keysInfo.length + " keys");
+
+    var active = 0;
+    var unactive = 0;
+    var fifteen = 0;
```
Comment (ametaireau) [inline tools/get_active_unactive_users.js]: biweekly ?

Comment (Natim) [inline tools/get_active_unactive_users.js]: http://www.wordreference.com/enfr/fortnight ?

Comment (leplatrem) [inline tools/get_active_unactive_users.js]: biweekly is more explicit for non native speakers... 
I saw twice-monthly too :)

File: tools/get_average_calls_per_user.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  var options = storage.settings;
+  var client = redis.createClient(
+    options.port,
+    options.host,
+    options.options
+  );
+  if (options.db) client.select(options.db);
+
+  var multi = client.multi();
+  multi.eval("return #redis.pcall('keys', 'userUrls.*')", 0);
```
Comment (ametaireau) [inline tools/get_average_calls_per_user.js]: wat wat wat

File: tools/utils.js [hunk 12]
```diff
@@ -0,0 +1,120 @@
+var fs = require('fs'),
+    rl = require('readline');
+
+function keysInformation(client, pattern, callback) {
```
Comment (ametaireau) [inline tools/utils.js]: Can we add some documentations here on what this does?

File: tools/utils.js [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+      if (key) {
+        multi.debug("object", key);
+      }
+
+    });
+    multi.exec(function(err, objects) {
+      if (err) return callback(err);
+      var results = [];
+
+      for(var i = 0; i < keys.length; i++) {
```
Comment (ametaireau) [inline tools/utils.js]: you can probably use `keys.forEach(function(key, i)` instead.

File: tools/utils.js [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+        });
+        results.push(data);
+      }
+
+      callback(null, results);
+    });
+  });
+}
+
+function dbInformation(client, db, callback) {
```
Comment (ametaireau) [inline tools/utils.js]: Same, please add a comment.

File: tools/get_average_calls_per_user.js (hunk 11 above)
Comment (Natim) [inline tools/get_average_calls_per_user.js]: It is some lua code that let you get the number of keys matching the pattern on server side.

**Comment by ametaireau** [comment]
Thanks, awesome!

**Pull request merged**